cozy/codegen/rpython.py

```diff
@@ -0,0 +1,422 @@
+            self.write("len(")
+            self.visit(node.e)
+            self.write(")")
+        elif op == "not":
+            self.write("not ")
+            self.visit(node.e)
+        elif op == "empty":
+            self.write("not bool(")
+            self.visit(node.e)
+            self.write(")")
+        elif op == "exists":
+            self.write("bool(")
+            self.visit(node.e)
+            self.write(")")
+        else:
+            import pdb; pdb.set_trace()
+            assert False, "Unhandled unary operation while writing RPython"
+
+    def visit_EBinOp(self, node):
+        # Rather than dispatching on op, it turns out to be simpler for us to
+        # dispatch by type!
+        if isinstance(node.type, (TInt, TBool)):
+            with self.parens():
+                self.visit(node.e1)
+                self.write(" ", node.op, " ")
+                self.visit(node.e2)
+        else:
+            import pdb; pdb.set_trace()
+            assert False, "Unhandled binary operation while writing RPython"
+
+    def visit_ECond(self, node):
+        # Quirk: Sometimes cond will be constantly true or false.
+        if isinstance(node, EBool):
+            self.visit(node.then_branch if node.val else node.else_branch)
+
+        with self.parens():
+            self.visit(node.then_branch)
+            self.write(" if ")
+            self.visit(node.cond)
+            self.write(" else ")
+            self.visit(node.else_branch)
+
+    def visit_EMap(self, node):
+        f = node.f
+        # Quirk: Sometimes we will see (map (\x -> x) xs) and we want to just
+        # have xs instead.
+        if f.arg == f.body:
+            self.visit(node.e)
+            return
+
+        self.write("[")
+        self.visit(f.body)
+        self.write(" for ")
+        # Hack: Write the lambda's arg as if it were an expr. This is close
+        # enough for now but definitely wrong. ~ C.
+        self.visit(f.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write("]")
+
+    def visit_EFilter(self, node):
+        p = node.p
+        self.write("[")
+        # Double hack! Abuse the arg again.
+        self.visit(p.arg)
+        self.write(" for ")
+        self.visit(p.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write(" if ")
+        self.visit(p.body)
+        self.write("]")
+
+    def visit_EArgMax(self, node):
+        f = node.f
+        self.needs["argMax"] = True
+        self.write("_argMax([")
+        self.visit(f.body)
+        self.write(" for ")
+        self.visit(f.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write("])")
+
+    def visit_EArgMin(self, node):
+        f = node.f
+        self.needs["argMin"] = True
+        self.write("_argMin([")
+        self.visit(f.body)
+        self.write(" for ")
+        self.visit(f.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write("])")
```

Comment:
The implementation for `argmin` and `argmax` here is not quite right. The Cozy expression `EArgMin(bag, f)` should translate to `min(bag, key=f) if bag else DEFAULT` (or the equivalent in whatever target language). The value DEFAULT is given by [`construct_value`](https://github.com/CozySynthesizer/cozy/blob/60e74947ce55ab0e6f0c76f7dd28dc3a6e8bbf68/cozy/evaluation.py#L80) for the bag's element type (currently awkwardly located in the `evaluation` module; it likely deserves a better home somewhere else).

(I have never written down a proper semantics for Cozy's intermediate language, so I'm actually quite impressed you were able to make any headway on this project at all!)

Reply:
Ah. I found a comment about bags and how all operations are total. With that in mind, I definitely need to look more closely at how lists are expected to work, too. The totality isn't hard to achieve with an RPython "bag", implemented as a dict mapping to ints, ala `collections.Counter`.

Reply:
(I have a background in applied reverse-engineering: GPUs, video games, and these days Somebody Else's Django code. Also I anticipated that this was going to be 2KLoC over two weekends, so I feel *great* so far.)